### ISSUE TITLE
Photos are always in landscape

### ISSUE THREAD

**Issue by cutiko** [fabian7593/MagicalCamera#7]
Greetings

I have being using this since a while now, and it is great. However, today I notice, always photos are stored in landscape orientation, what if the user take a photo in portrait orientation.

I thought this could be just my phone, so I try it with another, and same problem in both (Samsung galaxy note neo 3 and Sony xperia).

**Comment by fabian7593** [comment]
Hello, thanks for use MagicalCamera.

This issue is for any devices, because for example, the activity of camera of samsung's device is always in land scape, why? I don't know...

I know that for samsung devices, but not for sony, this is new for me, thanks for the issue in the new version of magical camera I fix it.

**Comment by cutiko** [comment]
@fabian7593 Hello, and thanks to you for such a magical library (xD pun intended)

I just noticed some similar issue in other application, so maybe it is me who does not know how to take photos. I'm very confused right now.

So please my most sincere excuses. However, I don't feel this issue is properly covered. Would you be so kind of allowing me to test it again and be back at you?

**Comment by fabian7593** [comment]
Yes @cutiko , I know why do you have this issue, these issues are because some devices have the Intent of camera in landscape, for this reason you have this issue in any applications... but not working same with all devices...
In the next version of magical camera I can fix it, and yes you can call me when you have another issue.

and again thanks for use my library.

**Comment by fabian7593** [comment]
Hello @cutiko see the new version 3.0 of magical camera, this have the fix of landscape photos...
in this week I add the functionality example in my readme.

**Comment by cutiko** [comment]
@fabian7593 Wow! I cant believe all the extras you add! This is great, thanks you so much. Maybe you could mark this as solved then.

**Comment by cutiko** [comment]
@fabian7593 I try the version 3.0 particularly the feature for rotating the images.

`magicalCamera.resultPhoto(requestCode, resultCode, data, true);`

I try with true at first and the change it to false. If is true, then the pictures are always in portrait, but if it is false then the pictures are always in landscape.

The expected result is if the picture was taken in landscape then the result should be in landscape, and viceversa.

I also try it without this new param, but takes me back to the original problem.

All attempts were made in Samsung Galaxy Note 3 neo and then replicated in Sony Xperia.

I dig a little further and notice the orientation is null `magicalCamera.getOrientation()`, maybe it is related...

**Comment by fabian7593** [comment]
@cutiko Thanks for report the issue, in the next version I can to fix it...
Otherwise,  the getOrientation() method only return a value when the user use before the getImageInformation() of the image... for get the photo information.

Thanks and sorry because I believe that I fix the orientation bug :s

**Comment by cutiko** [comment]
@fabian7593 I was using it inside the `getImageInformation()` validation, just as is done in the example. Also noticed that the `magicalCamera.getIso()`. The other image information returned a normal String. 

Since the null validation is check in the example I assumed was a known issue. This only happened with the Samsung but not with the Sony phone. This last lead me to believe it was a device related issue.

Thanks to you and dont worry, your library is great, does a lot of heavy lifting, Im just letting you know to try to help.

**Comment by fabian7593** [comment]
Hello again @cutiko, can you test the rotate image with the new methods for do this?
Now this method is open for user, and not own of the library... ç
Thanks

**Comment by cutiko** [comment]
@fabian7593 I did some testing, you can see it [in this gist](https://gist.github.com/cutiko/ce5abc9b0bb8fec22f20db10cd5fc266)

The rotation feature is indeed working, but there is a catch. The saved photo is not saved rotated. If the rotation could save the photo it would be great. You could do it automatically, it will save a new photo or it will override the previous photo. The other way is to allow us to concatenate some method .save() that resave the rotated photo. If a new photo is created we have to be able to get the path of the photo, so we can send it to any cloud storage (S3, FirebaseStorage).

I think you are moving on the right way with this feature. With the troublesome devices Im testing I have the same problems in other apps. So if the user can rotate the picture after taking it the problem is solved. There is no way this can be done without user intervention, considering that the orientation can be null and if is not null is not reliable. So after the user see the photo, then the user can rotate it, if is needed.

There is something else, you are still using MagicalCamera dot something as a param for the rotation, maybe for consistency everything should be MagicalCameraObject dot something (unless it breaks the poo encapsulation).

**Comment by fabian7593** [comment]
Yes you have completely reason, I achieve identify the device with this problem, because the camera hardware is put in landscape...
Sony, Samsung, and others brands, but not all the devices of this brand, and for me the best solution is leave it open to the user...
Yes and you have the posibility of test with magicalCamera.resultPhoto(requestCode, resultCode, data, MagicalCamera.ORIENTATION_ROTATE_180);, and the photo save with rotation.

And I check for the other version of magical camera the params with magicalcamera, or magicalCameraObject :S , Im sorry for this issue.

**Comment by cutiko** [comment]
I completely missed that method! That will fix it! Great, thanks, sorry
about that

**Comment by cutiko** [comment]
Greetings.

I did the testing. The problem here is, there is no way to know if the photo is right or wrong, unless the user visually confirm it. As you explained, there are other factors that make the photo being always in landscape. So there are two possibilities:

1. The user take the picture with the phone in portrait and is by default in landscape
2. The user take the photo in landscape and is correctly by default in landscape.

This is important because, only after the user take the photo, the user can see the result and then the user can choose if the photo is in the correct orientation.

Considering there is no reliable way to know the correct orientation then is the user the responsible for confirming the photo is right, and us the developers to provide the user with the UI and UX to do it so.

According to all above, this method `magicalCamera.resultPhoto(requestCode, resultCode, data, MagicalCamera.ORIENTATION_ROTATE_90);` is not useful. Because in the second case the photo will be rotated but it was correct (btw, the method is working).

So what we have to do is to allow the user to confirm the photo is right, and then save it. Or maybe save it, and then if the user rotate it, save it again. I figure a work arround:

```
    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        try {
            magicalCamera.resultPhoto(requestCode, resultCode, data);
        } catch (OutOfMemoryError error) {
            Toast.makeText(getContext(), "Out of Memory", Toast.LENGTH_SHORT).show();
        }

        final Bitmap bitmap = magicalCamera.getPhoto();

        //This can be null some times
        Toast.makeText(getContext(), String.valueOf(magicalCamera.getPrivateInformation().getOrientation()), Toast.LENGTH_SHORT).show();

        imageView.setImageBitmap(bitmap);
        button.setVisibility(View.VISIBLE);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //This will rotate the image, the problem is, the saved image is not rotated
                try {
                    Bitmap rotatedBitmap = magicalCamera.rotatePicture(bitmap, MagicalCamera.ORIENTATION_ROTATE_90);
                    imageView.setImageBitmap(rotatedBitmap);
                    //To be able to save the rotated photo the rotated bitmap must be obtained and pass as an argument to the save method
                    //As you can see this override the previous photo since the name is the same in both cases.
                    magicalCamera.savePhotoInMemoryDevice(rotatedBitmap, "rotatedPhoto", "myDirectoryName", MagicalCamera.JPEG, false);
                } catch (OutOfMemoryError e) {
                    Toast.makeText(getContext(), "Out Of Memory", Toast.LENGTH_SHORT).show();
                }
            }
        });

        magicalCamera.savePhotoInMemoryDevice(magicalCamera.getPhoto(), "rotatedPhoto", "myDirectoryName", MagicalCamera.JPEG, false);

    }
```

So what Im doing is to always save the photo in case the photo is in the right orientation. In case the user rotated the photo, the photo is saved again. The previous file is overridden since the name is the same. In other case developers could choose to create a new photo.

I want to do some recommendations:

1. Add some sort of explanation and code examples about this problem and how to solve it in the docs
2. Remove the `magicalCamera.resultPhoto(requestCode, resultCode, data, MagicalCamera.ORIENTATION_ROTATE_90);` method
3. In the future, could you consider a way the `magicalCamera.rotatePicture(bitmap, MagicalCamera.ORIENTATION_ROTATE_90);` method also save the rotated photo?

**As long as I concern, this is no longer an issue, I would be really happy to see it marked as resolved. Thanks for everything. I wish I could help more.**

**Comment by fabian7593** [comment]
@cutiko You have all reason of this, I have a new version of magical camera 5.0.2 ... and maybe a simple solution for image orientation is the user desition... please downalod the new example test and tell me how  thit it go.

**Comment by brspurri** [comment]
I'm running into the same issue. And I am using 5.0.2. 
Photos on some devices are landscape, while the same code results in portrait images on other devices.

**Comment by fabian7593** [comment]
Hello @brspurri , yes this is because any device hardware camera was made with camera in ladscape and others in portrait, I dont know why, but this problem even in mega apps (Fb, Insta, pinte... etc)...
my solution in this case is give to the user the posibillity of rotate image himselve, While I find a better solution... please download the example app of magical camera... and test in your device, I put and example of this here...
Thanks for use magical camera, please tell how you finished going.
